subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalker.java

```diff
@@ -38,47 +45,58 @@ private AbstractTypeMetadataWalker(TypeMetadataStore typeMetadataStore, Class<?
         } else if (Map.class.isAssignableFrom(nodeType) && !typeMetadata.hasAnnotatedProperties()) {
-            handleMap(node, (name, child) -> walk(child, getQualifiedName(qualifiedName, name), visitor));
+            handleMap(node, (name, child) -> walk(child, getQualifiedName(qualifiedName, name), visitor, previousNestedNodesWalkedOnPath));
         } else if (Iterable.class.isAssignableFrom(nodeType) && !typeMetadata.hasAnnotatedProperties()) {
-            handleIterable(node, (name, child) -> walk(child, getQualifiedName(qualifiedName, name), visitor));
+            handleIterable(node, (name, child) -> walk(child, getQualifiedName(qualifiedName, name), visitor, previousNestedNodesWalkedOnPath));
         } else {
-            visitor.visitNested(typeMetadata, qualifiedName, node);
-            typeMetadata.getPropertiesMetadata().forEach(propertyMetadata -> {
-                String childQualifiedName = getQualifiedName(qualifiedName, propertyMetadata.getPropertyName());
-                if (propertyMetadata.getPropertyType() == nestedAnnotation) {
-                    T child = getChild(node, propertyMetadata);
-                    walk(child, childQualifiedName, visitor);
-                } else {
-                    visitor.visitLeaf(childQualifiedName, propertyMetadata, () -> getChild(node, propertyMetadata));
-                }
-            });
+            handleNested(node, typeMetadata, qualifiedName, visitor, previousNestedNodesWalkedOnPath);
         }
     }
 
-    private static String getQualifiedName(@Nullable String parentPropertyName, String childPropertyName) {
-        return parentPropertyName == null
-            ? childPropertyName
-            : parentPropertyName + "." + childPropertyName;
+    private void handleNested(T node, TypeMetadata typeMetadata, @Nullable String qualifiedName, NodeMetadataVisitor<T> visitor, Set<T> previousNestedNodesOnPath) {
+        if (previousNestedNodesOnPath.contains(node)) {
+            return;
+        }
+
+        visitor.visitNested(typeMetadata, qualifiedName, node);
+        Set<T> nestedNodesOnPath = newIdentitySetOf(previousNestedNodesOnPath, node);
+        typeMetadata.getPropertiesMetadata().forEach(propertyMetadata -> {
+            String childQualifiedName = getQualifiedName(qualifiedName, propertyMetadata.getPropertyName());
+            if (propertyMetadata.getPropertyType() == nestedAnnotation) {
+                Optional<T> childOptional = getChild(node, propertyMetadata);
+                childOptional.ifPresent(child -> walk(child, childQualifiedName, visitor, nestedNodesOnPath));
+            } else {
+                visitor.visitLeaf(childQualifiedName, propertyMetadata, () -> getChild(node, propertyMetadata).orElse(null));
+            }
+        });
     }
 
-    protected abstract void handleProvider(T node, Consumer<T> handler);
+    abstract void handleProvider(T node, Consumer<T> handler);
```

Comment:
Why remove `protected` here?

Reply:
So it's consistent with other methods, also it seems we don't need `protected` here anyway and package-private should do it.